### PR TITLE
Feature toggles

### PR DESCRIPTION
**Pull request by chris-hinds**
Does not resolve but starts https://github.com/bbc/simorgh/issues/2234 

**Overall change:** Add a feature toggle service for global application toggles. The intention is to expand this later with the ability to update feature toggles from a remote endpoint hence the inclusion of an action which can be dispatched to update the context value for the feature toggles.

**Code changes:**

- Add's a toggle context
- Add's a custom hook for returning the value of a toggle
- Add's a new container for conditionally rendering child components or fallback UI based on the value of a toggle.
- Adds a new withContext HOC to remove the context providers from the pageWrapper (this can now just be related to page furniture and not context providers.)
---

- [X] I have assigned myself to this PR and the corresponding issues
- [X] Tests added for new features
- [ ] Test engineer approval

### CODE REVIEW

File: src/app/containers/Toggle/README.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+## Overview
+
+The toggle container is a component that is designed to wrap other components and render the child component based on a feature toggle.
+
+## Feature Toggles
+
+At the time of writing feature toggles are designed to be global (i.e not service specific) and are dependant on the application environment.
+feature toggles can be found in `src/app/lib/config/toggles`
+
+## Useage
```
Comment (jroebu14) [inline src/app/containers/Toggle/README.md]: Typo here. Should be "Usage"

File: src/app/containers/Toggle/README.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    <ChildComponent>
+        <p>I will be rendered if the enabled value of toggle: foo is set to true</p>
+    </ChildComponent>
+</Toggle>
+```
+
+### Rendering a fallback UI component if the toggle value is set to false
+
+```
+<Toggle toggleName="foo" fallbackComponent={<h1>I am the fallback UI</h1>}>
```
Comment (jroebu14) [inline src/app/containers/Toggle/README.md]: Just wondering if a render prop component is maybe a more appropriate API here?

Also, why provide 2 methods of using toggles? When would we use one over the other? Could we get away with using only the `useToggle` hook?

File: src/app/containers/Toggle/toggleUtils.js [hunk 3]
```diff
@@ -0,0 +1,7 @@
+const isToggleEnabled = (toggleName, toggles) => {
```
Comment (jroebu14) [inline src/app/containers/Toggle/toggleUtils.js]: Could be more concise e.g.

```jsx
const isToggleEnabled = (toggleName, toggles) => Boolean(toggles[toggleName])
```

File: src/app/containers/Toggle/toggleUtils.test.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import isToggleEnabled from './toggleUtils';
```
Comment (jroebu14) [inline src/app/containers/Toggle/toggleUtils.test.js]: Might there be some confusion that this is not imported like

```js
import { isToggleEnabled } from './toggleUtils';
```

File: src/app/containers/Toggle/README.md (hunk 1 above)
Comment (chris-hinds) [reply src/app/containers/Toggle/README.md]: Good spot, fixing now.

File: src/app/containers/Toggle/README.md (hunk 2 above)
Comment (chris-hinds) [reply src/app/containers/Toggle/README.md]: The reason for their being two approaches is that in some scenarios I can see time where we might not want to wrap a component to out it behind a feature toggle. But rather just grab the boolean value of the toggle and let the component handle the conditional rendering.

Where as sometimes the wrapping component would prove a more simple/readable method of using a feature toggle and in my eyes a better way to pass in a fallback UI component.

Either way the wrapping container uses the useToggle hook.

File: src/app/containers/Toggle/toggleUtils.test.js (hunk 4 above)
Comment (chris-hinds) [reply src/app/containers/Toggle/toggleUtils.test.js]: could import it that that way, however eslint then complains about preferring a default export. I am happy either way.

File: src/app/containers/Toggle/toggleUtils.js (hunk 3 above)
Comment (chris-hinds) [reply src/app/containers/Toggle/toggleUtils.js]: So the remote toggles endpoint we would eventually use, returns a object for each toggle, 

```
toggle: {
    enabled: true,
    value: 'some string'
}
```

The current method here return the entire object and the useToggle hook just pulls the `enabled` value form it. In the future we may well want to also use the `value` key.

So always returning a boolean here would not allow that.

File: src/app/containers/Toggle/toggleUtils.test.js (hunk 4 above)
Comment (jroebu14) [reply src/app/containers/Toggle/toggleUtils.test.js]: Sorry for being pernickety! What about renaming the file to `isToggleEnabled.js` and then in the future if we have more than one toggle utility function we could have a directory called `toggleUtils` with an index that re-exports the utility functions.

File: src/app/Layouts/defaultPageWrapper.jsx [hunk 5]
```diff
@@ -1,60 +1,26 @@
...
-        <ManifestContainer />
-        <ConsentBanner />
-        <HeaderContainer />
-        {children}
-        <FooterContainer />
-      </RequestContextProvider>
-    </ServiceContextProvider>
-  </DialContextProvider>
+const PageWrapper = ({ children }) => (
+  <>
```
Comment (dr3) [inline src/app/Layouts/defaultPageWrapper.jsx]: Hey, i dont think we should mix our usage of `<Fragment>` and `<>` in simorgh, can this be `<Fragment>` for now, and raise an issue to make all uses the same across simorgh, if you want us to change to it please?

**Review by dr3 — CHANGES_REQUESTED** [review]

File: src/app/Layouts/defaultPageWrapper.jsx (hunk 5 above)
Comment (chris-hinds) [reply src/app/Layouts/defaultPageWrapper.jsx]: Updated.

File: src/app/containers/Toggle/README.md (hunk 2 above)
Comment (jroebu14) [reply src/app/containers/Toggle/README.md]: Ah ok I see.

Should the `fallbackComponent` prop here be `FallbackComponent`? Also, should the example with `fallbackComponent={<h1>I am the fallback UI</h1>}>` not be

```jsx
const FallbackUI = () => (<h1>I am the fallback UI</h1>)

<Toggle toggleName="foo" FallbackComponent={FallbackUI}>
```

and especially not

```jsx
fallbackComponent={<FallbackUI />}>
```

File: src/app/containers/Toggle/README.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
+# Toggle Container
+
+## Overview
+
+The toggle container is a component that is designed to wrap other components and render the child component based on a feature toggle.
+
+## Feature Toggles
+
+At the time of writing feature toggles are designed to be global (i.e not service specific) and are dependant on the application environment.
+feature toggles can be found in `src/app/lib/config/toggles`
```
Comment (dr3) [inline src/app/containers/Toggle/README.md]:
```suggestion
Feature toggles can be found in `src/app/lib/config/toggles`.
```

File: src/app/containers/Toggle/index.jsx [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+import useToggle from './useToggle';
+
+const ToggleContainer = ({ toggleName, FallbackComponent, children }) => {
+  const { enabled } = useToggle(toggleName);
+
+  if (enabled) {
+    return children;
+  }
+
+  if (FallbackComponent) {
```
Comment (dr3) [inline src/app/containers/Toggle/index.jsx:12]: technically you could just return `FallbackComponent` without the `if` here, as it default to null, but feel free to leave it

File: src/app/containers/Toggle/toggleUtils.test.js (hunk 4 above)
Comment (dr3) [reply src/app/containers/Toggle/toggleUtils.test.js]: Yeah I think either override the eslint rule and export as `{ isToggleEnabled }` or make a `utils` folder and add a `isToggleEnabled.js` (which i prefer)

File: src/app/reducers/ToggleReducer/index.jsx [hunk 8]
```diff
@@ -0,0 +1,19 @@
+// toggle actions
```
Comment (dr3) [inline src/app/reducers/ToggleReducer/index.jsx:1]: whats this?

**Review by dr3 — CHANGES_REQUESTED** [review]

File: src/app/containers/Toggle/index.jsx (hunk 7 above)
Comment (chris-hinds) [reply src/app/containers/Toggle/index.jsx:12]: I think wrapping it in the if statement is a bit more descriptive on what is actually happening.

File: src/app/reducers/ToggleReducer/index.jsx (hunk 8 above)
Comment (chris-hinds) [reply src/app/reducers/ToggleReducer/index.jsx:1]: So this is an action which can be dispatched based on some event, e.g a user interaction, a react lifecycle a fetch event etc.

This is following a common pattern used when implementing state management, where an interaction should not directly modify the state. Instead it dispatches an action with a data payload. This action then triggers the reducer and it is the reducer that actually handles the updating of state.

I have left this in as this action will be used when we allow feature toggles to be updated client side.

File: src/app/containers/Toggle/README.md (hunk 6 above)
Comment (chris-hinds) [reply src/app/containers/Toggle/README.md]: Have updated this, thanks.

**Review by dr3 — APPROVED** [review]

**Comment by dr3** [comment]
Will require regression of existing pages due to changes to page wrappers. Ensure all contexts work as expected

File: src/app/containers/Toggle/README.md (hunk 2 above)
Comment (jroebu14) [reply src/app/containers/Toggle/README.md]: So would `<Toggle toggleName="foo" fallbackComponent={<h1>I am the fallback UI</h1>}>` actually work?

@hindsc52

**Review by jroebu14 — APPROVED** [review]

**Comment by PriyaKR** [comment]
Testing:Looks good to me.

Tested for timestamp and caption container by wrapping them within `<Toggle>....</Toggle>` tags and enabling or disabling the toggle within ``/src/app/lib/config/toggles``.And also disabling the feature and providing a fallback works fine.

**Comment by PriyaKR** [comment]
Good to be merged.

**Comment by chris-hinds** [comment]
We can ignore these CC issues as CC is being very restrictive for a react functional component, especially considering our eslint rules

**Pull request merged**